cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+      }
+      ARROW_ASSIGN_OR_RAISE(ExecBatch eb, ExecBatch::Make(std::move(partsupp_result)));
+      partsupp_results.emplace_back(std::move(eb));
+    }
+    // Return the first batch, enqueue the rest.
+    {
+      std::lock_guard<std::mutex> lock(partsupp_output_queue_mutex_);
+      for (size_t i = 1; i < partsupp_results.size(); i++)
+        partsupp_output_queue_.emplace(std::move(partsupp_results[i]));
+    }
+    return std::move(partsupp_results[0]);
+  }
+
+ private:
+#define FOR_EACH_PART_COLUMN(F) \
+  F(P_PARTKEY)                  \
+  F(P_NAME)                     \
+  F(P_MFGR)                     \
+  F(P_BRAND)                    \
+  F(P_TYPE)                     \
+  F(P_SIZE)                     \
+  F(P_CONTAINER)                \
+  F(P_RETAILPRICE)              \
+  F(P_COMMENT)
+
+#define FOR_EACH_PARTSUPP_COLUMN(F) \
+  F(PS_PARTKEY)                     \
+  F(PS_SUPPKEY)                     \
+  F(PS_AVAILQTY)                    \
+  F(PS_SUPPLYCOST)                  \
+  F(PS_COMMENT)
+
+#define MAKE_ENUM(col) col,
+  struct PART {
+    enum {
+      FOR_EACH_PART_COLUMN(MAKE_ENUM) kNumCols,
+    };
+  };
+  struct PARTSUPP {
+    enum {
+      FOR_EACH_PARTSUPP_COLUMN(MAKE_ENUM) kNumCols,
+    };
+  };
+
+#define MAKE_STRING_MAP(col) {#col, PART::col},
+  const std::unordered_map<std::string, int> kPartNameMap = {
+      FOR_EACH_PART_COLUMN(MAKE_STRING_MAP)};
+#undef MAKE_STRING_MAP
+#define MAKE_STRING_MAP(col) {#col, PARTSUPP::col},
+  const std::unordered_map<std::string, int> kPartsuppNameMap = {
+      FOR_EACH_PARTSUPP_COLUMN(MAKE_STRING_MAP)};
+#undef MAKE_STRING_MAP
+#define MAKE_FN_ARRAY(col) \
+  [this](size_t thread_index) { return this->col(thread_index); },
+  std::vector<GenerateColumnFn> kPartGenerators = {FOR_EACH_PART_COLUMN(MAKE_FN_ARRAY)};
+  std::vector<GenerateColumnFn> kPartsuppGenerators = {
+      FOR_EACH_PARTSUPP_COLUMN(MAKE_FN_ARRAY)};
+#undef MAKE_FN_ARRAY
+#undef FOR_EACH_LINEITEM_COLUMN
+#undef FOR_EACH_ORDERS_COLUMN
+
+  const std::vector<std::shared_ptr<DataType>> kPartTypes = {
+      int32(), utf8(),  fixed_size_binary(25), fixed_size_binary(10),
+      utf8(),  int32(), fixed_size_binary(10), decimal(12, 2),
+      utf8(),
+  };
+
+  const std::vector<std::shared_ptr<DataType>> kPartsuppTypes = {
+      int32(), int32(), int32(), decimal(12, 2), utf8(),
+  };
+
+  Status AllocatePartBatch(size_t thread_index, int column) {
+    ThreadLocalData& tld = thread_local_data_[thread_index];
+    ARROW_DCHECK(tld.part[column].kind() == Datum::NONE);
+    int32_t byte_width = arrow::internal::GetByteWidth(*kPartTypes[column]);
+    ARROW_ASSIGN_OR_RAISE(std::unique_ptr<Buffer> buff,
+                          AllocateBuffer(tld.part_to_generate * byte_width));
+    ArrayData ad(kPartTypes[column], tld.part_to_generate, {nullptr, std::move(buff)});
+    tld.part[column] = std::move(ad);
+    return Status::OK();
+  }
+
+  Status P_PARTKEY(size_t thread_index) {
```

Comment:
Can you add comments describing and explaining the various generation methods below?
We should be able to understand and maintain this code without knowing the TPC-H spec by heart, ideally :-)

Reply:
i added a big one at the top